### ISSUE TITLE
findOrCreate(): Populates returned items

### ISSUE THREAD

**Issue by weyert** [balderdashy/waterline#1211]
The code below finds the record correctly but also populates the association of this required while I need request this. Why does this happen?

    User
      .findOrCreate({facebookId: 'id'}, {email:'me@email.com, facebookId: 'id'})
      .then(user => {
        if (!user) {
          return next(null, null, sails.config.errors.AUTH_FAILED);
        }
        return next(null, user, {});
      })

Where can I find the logic for findOrCreate() in Waterline? I only seem to find findOrCreateEach() when searching for it.

**Comment by devinivy** [comment]
What is being populated exactly?  I'm a little bit confused what issue you're seeing.  The logic for that method can be found here: https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/composite.js#L24

**Comment by weyert** [comment]
Sorry, missed your message. It's populating the `notifications`-association in the following Model definition. I don't understand why as I didn't call populate('notifications').

import _ from 'lodash';

/**
 * User
 * @description :: Model for storing users
 */

export default {
  schema: true,

  protectedAttributes: function () {
    return [
      'totalNotifications',
      'facebookId', 'facebookAccessToken'
    ];
  },

  attributes: {

    username: {
      type: 'string',
      required: false,
      unique: true,
      alphanumericdashed: true
    },

    password: {
      type: 'string'
    },

    totalNotifications: {
      type: 'integer',
      columnName: 'notifications_count',
      defaultsTo: 0
    },
    notifications: {
      collection: 'Notification',
      via: 'creator'
    },

    /**
     * Social network integration
     */
    facebookId: {
      type: 'string',
      columnName: 'facebook_id',
      defaultsTo: ''
    },

    facebookAccessToken: {
      type: 'string',
      columnName: 'facebook_token',
      defaultsTo: ''
    },

    /**
     * Returns the full name of the user
     *
     * @returns {string}
     */
    fullName: function() {
      return this.firstName + ' ' + this.lastName
    },

    /**
     * Returns the full name of the user
     *
     * @returns {string}
     */
    shortName: function() {
      var result = this.firstName;
      if (this.lastName && this.lastName.length > 2) {
        result += ' ' + this.lastName.charAt(0).toUpperCase();
      }

      return result;
    },
};

**Comment by jvanalst** [comment]
Actually, that's my question as well.

By default .findOrCreate() seems to populate all the associations, and this can be a big overhead if you have one-to-many relationship. If I'm not mistake that's also the reverse of the behaviour from the .find() method, which populates nothing until you call .populate().

Is there anyway to turn off this population on findOrCreate?

**Comment by particlebanana** [comment]
How are you calling `findOrCreate`? There is no code in that method that would populate anything and nothing in my test that shows it's being populated?

**Comment by weyert** [comment]
@particlebanana, yes, that's true, that's the whole problem. Somehow it's populating while I haven't even directly requested it to do by calling `populate()`. I am calling `findOrCreate()` as shown in my first comment in this thread. Only happens for me when using `findOrCreate()` with `find` or `findOne` it works fine.

**Comment by particlebanana** [comment]
@weyert can you isolate this into something that can be reproduced? Maybe post a project I can clone and look at? I am unable to recreate this using `findOrCreate`.

**Comment by weyert** [comment]
I will see what I can do. I am using the postgresql adapter for Waterline.

**Comment by tobalsgithub** [comment]
We're also using the postgresql adapter for Waterline and seeing the same problem. All associations get populated when `findOrCreate` is called. More specifically for us, we noticed an issue when calling `Model.update(id, record)` where `record` had some associations populated. Since this function uses `findOrCreate`, it was populating every association, which for us was very costly. 

Maybe a problem in the postgresql adapter? I can start digging there.

**Comment by particlebanana** [comment]
Ah this is easier to understand then. Whenever you send a record that has populations to an `update` or `create` method they will get "synced" which essentially means it will do a deep update or deep create. This is probably not what you want! Looking back we should have wrapped all this logic in a flag of some sort so it can be toggled.

For now you want to make sure you are sending values in that haven't been populated. In the next major version the nested updates will be either removed or wrapped in a flag.

**Comment by tobalsgithub** [comment]
Yeah, we updated our code so we're not doing this anymore.

I'm curious though... you say that it will do a deep `update` or deep `create`, which I understand. What I don't understand is that everything is getting populated, even associations that were not populated or present at all. Is this still what you would expect?

**Comment by particlebanana** [comment]
If the value is present in the values and it's an association it **should** get flagged as changed and set to populate. In reality that stuff is pretty messy and definitely needs some TLC.

Our [test fixtures](https://github.com/balderdashy/waterline-adapter-tests/blob/master/interfaces/associations/belongsTo/update.nested.js) for that don't seem to have more than 1 association on them so that would be a good check to narrow down the behavior.

**Comment by tobalsgithub** [comment]
Not sure you quite got my questions. Let's say we have 3 models, Person, Cars and Pets. 

A Person can have many cars.
A Person can have only 1 pet. 

If we have an instance of a person model where `pet` is populated but `cars` is not:  

```
person: {
  id: 4,
  pet: {
    id: 123,
    name: 'Dog'
  }
}
```

and we call `Person.update(4, person, function...)`, both `person.pet` and `person.cars` get populated. 

That's what I don't understand. Certainly I understand why `pet` gets populated. But why would `cars` get populated?

**Comment by particlebanana** [comment]
Yeah we don't have a test for that case so we should add it and track down the behavior that is causing that. Mind submitting a PR to [waterline-adapter-tests](https://github.com/balderdashy/waterline-adapter-tests) that checks for this exact case?

**Comment by jvanalst** [comment]
Yup. @tobalsgithub is correct. This seems to be very similar to my situation. It's not isolated to the postgresql adaptor either, we've been using sails-disk and sails-mysql.

I'm doing a Person.findOrCreate and a Person.update with none of the associations populated in the update information, but it's still populating all them in the return value creating massive overhead (in my applications a person could potentially collect thousands of many-to-one records).

**Comment by tobalsgithub** [comment]
@particlebanana I can give it a whirl. Probably after thanksgiving though

**Comment by weyert** [comment]
Great, thanks :)

**Comment by sailsbot** [comment]
Thanks for posting, @weyert.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
